src/storage/mvcc/txn.rs

```diff
@@ -262,6 +273,19 @@ impl<S: Snapshot> MvccTxn<S> {
         let ts = self.start_ts;
         self.put_write(key, ts, write.to_bytes());
         self.unlock_key(key.clone());
+        if self.collapse_rollback {
+            self.collapse_prev_rollback(key)?;
+        }
+        Ok(())
+    }
+
+    fn collapse_prev_rollback(&mut self, key: &Key) -> Result<()> {
+        if let Some((commit_ts, write)) = self.reader.seek_write(key, u64::max_value())? {
+            assert!(commit_ts < self.start_ts);
```

Comment:
I think `commit_ts` could be greater than `self.start_ts`.
Consider txn1(ts=100) contains 2 keys [k1, k2], while k1 already has a rollback with higher version (ts=200).
When txn1 commits, request to lock k1 is delayed. After another txn2 meets the lock on k2, it will try to rollback txn1 by sending a rollback(k1, ts=100) to tikv.
I suppose we should check this case at  https://github.com/pingcap/tikv/pull/3290/files#diff-10ac7ebd80ca4da7cccc8e3539f12eb4R262 and just write nothing and return OK.

Reply:
great, we will TLA+ and find that we must consider the case `commit_ts > self.start_ts`, otherwise, the invariance check will fail.

Reply:
@disksing You said k1 already has a rollback with ts = 200, so before you call `collapse_pre_rollback`, there must be a related lock in k1 whose start_ts is larger than 200.

Reply:
@disksing I decided `collapse_pre_rollback` only works when current Rollback will be the latest commit version, but you remind me `collapse_pre_rollback` can be better. Thanks